library/proc_macro/src/diagnostic.rs

```diff
@@ -44,7 +48,22 @@ impl<'a> MultiSpan for &'a [Span] {
 }
 
 /// A structure representing a diagnostic message and associated children
-/// messages.
+/// messages. Diagnostics are used to report errors when a procedural macro
+/// is generating code, as it's more friendly to macro users and macro developers
+/// than panicking.
+/// # Examples
+/// Emitting a simple error diagnostic at the callsite.
+/// ```rust
+/// Diagnostic::spanned(Level::Error, "unclosed tag", Span::call_site())
```

Comment:
This API is likely to change soon with https://github.com/rust-lang/rust/issues/54140#issuecomment-1487529341. Not sure whether it's worth adding nice examples here in the meantime, but probably doesn't hurt.

Reply:
Thanks for input! I'll add your idea of `compile_error!` when I get home and work on the third commit of this PR. I'll add examples anyways as I might as well while I'm doing this. I can do another documentation PR when there's the new API.

Reply:
I might also play around with adding a feature gate as something like `proc_macro_diagnostics_new_api`, although I don't know if this would require something like an RFC to test out estebank's proposed API.